File: backups/python-legacy-20260112/scripts/discovery/discover-skills.py
#!/usr/bin/env python3
"""
Discover and index skills from multiple sources into hybrid-registry.json.

Scans marketplace, global, and project sources to build a unified skill index.
Detects local dependencies and resolves which source wins per skill.

Usage:
    python discover-skills.py [--verbose]
"""

import hashlib
import json
import os
import re
import sys
from datetime import datetime
from pathlib import Path

# Try to import yaml
try:
    import yaml
    HAS_YAML = True
except ImportError:
    HAS_YAML = False

# Configuration
CLAUDE_HOME = Path.home() / ".claude"
SCRIPT_DIR = Path(__file__).parent.parent  # Marketplace root

REGISTRY_FILE = SCRIPT_DIR / "configs" / "hybrid-registry.json"
PROJECTS_REGISTRY = SCRIPT_DIR / "configs" / "projects-registry.json"
SYNC_CONFIG = SCRIPT_DIR / "configs" / "sync-config.json"

VERBOSE = "--verbose" in sys.argv or "-v" in sys.argv


def log(msg: str, always: bool = False):
    """Print message if verbose or always."""
    if VERBOSE or always:
        print(msg)


def extract_yaml_manual(content: str) -> dict:
    """Manual YAML parsing for simple frontmatter."""
    match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return {}

    yaml_content = match.group(1)
    result = {}
    current_list = None

    for line in yaml_content.split('\n'):
        if not line.strip():
            continue

        if line.startswith('  - '):
            if current_list is not None:
                value = line[4:].strip().strip('"').strip("'")
                current_list.append(value)
            continue

        if ':' in line and not line.startswith(' '):
            key, _, value = line.partition(':')
            key = key.strip()
            value = value.strip()

            if value:
                result[key] = value.strip('"').strip("'")
                current_list = None
            else:
                result[key] = []
                current_list = result[key]

    return result


def extract_yaml(content: str) -> dict:
    """Extract YAML frontmatter from markdown content."""
    match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return {}

    if HAS_YAML:
        try:
            return yaml.safe_load(match.group(1)) or {}
        except yaml.YAMLError:
            return extract_yaml_manual(content)
    return extract_yaml_manual(content)


def extract_content_summary(content: str, max_length: int = 500) -> str:
    """Extract a summary from SKILL.md content."""
    content_without_fm = re.sub(r'^---\n.*?\n---\n?', '', content, flags=re.DOTALL)
    content_clean = re.sub(r'```[\s\S]*?```', '', content_without_fm)
    content_clean = re.sub(r'`[^`]+`', '', content_clean)

    headers = re.findall(r'^##+ (.+)$', content_clean, re.MULTILINE)
    paragraphs = [p.strip() for p in content_clean.split('\n\n') if p.strip()]
    first_para = paragraphs[0] if paragraphs else ""

    summary_parts = []
    if headers:
        meaningful = [h for h in headers if len(h) > 3 and h.lower() not in
                      ('usage', 'installation', 'configuration', 'examples', 'notes')]
        if meaningful:
            summary_parts.append(' | '.join(meaningful[:5]))

    if first_para:
        first_para = re.sub(r'\[([^\]]+)\]\([^)]+\)', r'\1', first_para)
        first_para = re.sub(r'[*_]{1,2}([^*_]+)[*_]{1,2}', r'\1', first_para)
        summary_parts.append(first_para[:300])

    summary = ' '.join(summary_parts)
    return summary[:max_length] if summary else ""


def detect_dependencies(content: str, skill_path: Path) -> dict:
    """Detect local dependencies in skill content."""
    deps = {
        "local_files": [],
        "requires_project_context": False,
        "external_scripts": []
    }

    # Find relative path references (../)
    relative_refs = re.findall(r'\.\./[^\s\)\]"\']+', content)
    if relative_refs:
        deps["local_files"] = list(set(relative_refs))
        deps["requires_project_context"] = True

    # Find script references
    script_refs = re.findall(r'scripts/[^\s\)\]"\']+\.py', content)
    if script_refs:
        deps["external_scripts"] = list(set(script_refs))

    return deps


def compute_hash(content: str) -> str:
    """Compute SHA256 hash of content."""
    return hashlib.sha256(content.encode('utf-8')).hexdigest()[:16]


def expand_path(path: str) -> Path:
    """Expand ~ and resolve path."""
    return Path(os.path.expanduser(path)).resolve()


def load_registry() -> dict:
    """Load or create hybrid registry."""
    if REGISTRY_FILE.exists():
        try:
            with open(REGISTRY_FILE, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            pass

    return {
        "version": "1.0.0",
        "description": "Hybrid skill registry",
        "last_indexed": None,
        "sources": [],
        "project_sources": [],
        "skills": {}
    }


def load_projects_registry() -> dict:
    """Load projects registry to find project paths."""
    if PROJECTS_REGISTRY.exists():
        try:
            with open(PROJECTS_REGISTRY, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            pass
    return {"projects": {}}


def load_sync_config() -> dict:
    """Load sync configuration."""
    if SYNC_CONFIG.exists():
        try:
            with open(SYNC_CONFIG, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (json.JSONDecodeError, IOError):
            pass
    return {}


def find_project_skill_sources(projects_reg: dict) -> list:
    """Find project sources based on project_sources patterns from sync-config.json."""
    sources = []
    sync_config = load_sync_config()

    for ps in sync_config.get("project_sources", []):
        pattern = ps.get("project_pattern", "")
        skills_path = ps.get("skills_path", ".claude/skills")
        priority = ps.get("priority", 2)

        # Search in projects registry
        for encoded_path, project in projects_reg.get("projects", {}).items():
            project_path = project.get("path", "")
            if not project_path:
                continue

            # Match pattern against project path
            import fnmatch
            if fnmatch.fnmatch(project_path, pattern) or pattern.replace("**/", "") in project_path:
                full_skills_path = Path(project_path) / skills_path
                if full_skills_path.exists():
                    source_id = f"project:{project.get('name', encoded_path)}"
                    sources.append({
                        "id": source_id,
                        "type": "project",
                        "path": str(full_skills_path),
                        "project_path": project_path,
                        "priority": priority,
                        "skills_pattern": "*/SKILL.md"
                    })
                    log(f"  Found project source: {source_id}")

    return sources


def scan_source(source: dict) -> list:
    """Scan a source for skills."""
    skills = []
    source_path = expand_path(source["path"])
    pattern = source.get("skills_pattern", "*/SKILL.md")

    if not source_path.exists():
        log(f"  Source path does not exist: {source_path}")
        return skills

    for skill_md in sorted(source_path.glob(pattern)):
        try:
            content = skill_md.read_text(encoding='utf-8')
            data = extract_yaml(content)

            name = data.get("name", skill_md.parent.name)
            description = data.get("description", "")
            triggers = data.get("triggers", [])

            if not triggers or not isinstance(triggers, list):
                log(f"    Skipped {name}: no triggers")
                continue

            content_summary = extract_content_summary(content)
            dependencies = detect_dependencies(content, skill_md)
            file_hash = compute_hash(content)
            mtime = datetime.fromtimestamp(skill_md.stat().st_mtime).isoformat()

            skill_info = {
                "name": name,
                "description": description,
                "triggers": triggers,
                "content_summary": content_summary,
                "location": {
                    "source": source["id"],
                    "path": str(skill_md.relative_to(source_path)),
                    "full_path": str(skill_md),
                    "hash": file_hash,
                    "mtime": mtime
                },
                "dependencies": dependencies
            }

            skills.append(skill_info)
            log(f"    Found: {name} ({len(triggers)} triggers)")

        except Exception as e:
            log(f"    Error processing {skill_md}: {e}")

    return skills


def resolve_skills(all_skills: list, sources: list) -> dict:
    """Resolve which source wins for each skill."""
    # Build priority map
    priority_map = {s["id"]: s.get("priority", 0) for s in sources}

    # Group by skill name
    by_name = {}
    for skill in all_skills:
        name = skill["name"]
        if name not in by_name:
            by_name[name] = []
        by_name[name].append(skill)

    # Resolve each skill
    resolved = {}
    for name, variants in by_name.items():
        # Sort by priority (higher wins)
        variants.sort(key=lambda s: priority_map.get(s["location"]["source"], 0), reverse=True)
        winner = variants[0]

        resolved[name] = {
            "name": name,
            "description": winner["description"],
            "triggers": winner["triggers"],
            "content_summary": winner.get("content_summary", ""),
            "locations": [v["location"] for v in variants],
            "resolved_source": winner["location"]["source"],
            "dependencies": winner.get("dependencies", {}),
            "scope": "project" if winner["location"]["source"].startswith("project:") else "global"
        }

        if len(variants) > 1:
            log(f"  Resolved {name}: {winner['location']['source']} wins over {[v['location']['source'] for v in variants[1:]]}")

    return resolved


def main():
    print("=" * 60)
    print("Hybrid Skill Discovery")
    print("=" * 60)

    # Load registries
    registry = load_registry()
    projects_reg = load_projects_registry()

    # Build sources list
    all_sources = list(registry.get("sources", []))

    # Add project sources (from sync-config.json)
    print("\nFinding project sources...")
    project_sources = find_project_skill_sources(projects_reg)
    all_sources.extend(project_sources)

    print(f"\nSources to scan: {len(all_sources)}")
    for s in all_sources:
        print(f"  [{s.get('priority', 0)}] {s['id']}: {s['path']}")

    # Scan all sources
    all_skills = []
    for source in all_sources:
        print(f"\nScanning {source['id']}...")
        skills = scan_source(source)
        all_skills.extend(skills)
        print(f"  Found {len(skills)} skills")

    # Resolve conflicts
    print("\nResolving skill priorities...")
    resolved = resolve_skills(all_skills, all_sources)

    # Update registry
    registry["skills"] = resolved
    registry["last_indexed"] = datetime.now().isoformat()

    # Save
    REGISTRY_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(REGISTRY_FILE, 'w', encoding='utf-8') as f:
        json.dump(registry, f, indent=2, ensure_ascii=False)

    # Summary
    print("\n" + "=" * 60)
    print("Summary")
    print("=" * 60)
    print(f"Total skills indexed: {len(resolved)}")

    # Count by scope
    global_count = sum(1 for s in resolved.values() if s["scope"] == "global")
    project_count = sum(1 for s in resolved.values() if s["scope"] == "project")
    print(f"  Global skills: {global_count}")
    print(f"  Project skills: {project_count}")

    # Count with dependencies
    with_deps = sum(1 for s in resolved.values() if s.get("dependencies", {}).get("requires_project_context"))
    print(f"  Skills with local dependencies: {with_deps}")

    # Multi-location skills
    multi = sum(1 for s in resolved.values() if len(s.get("locations", [])) > 1)
    print(f"  Skills in multiple locations: {multi}")

    print(f"\nRegistry saved to: {REGISTRY_FILE}")


if __name__ == "__main__":
    main()

File: scripts/discovery/discover-skills.js
#!/usr/bin/env node
/**
 * Discover Skills - Build hybrid registry from all sources.
 *
 * Scans marketplace, global (~/.claude/skills), and project sources
 * to build a unified hybrid registry with priority resolution.
 *
 * Pattern from discover-skills.py
 */

const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const { program } = require('commander');
const fg = require('fast-glob');

const { extractYaml, extractContentSummary } = require('../lib/yaml-parser');
const { readJson, writeJson, computeHash, expandPath, fileExists, readFile } = require('../lib/file-utils');

// Constants
const MARKETPLACE_ROOT = path.resolve(__dirname, '../..');
const CLAUDE_HOME = path.join(os.homedir(), '.claude');
const REGISTRY_DIR = path.join(MARKETPLACE_ROOT, 'registry');
const SYNC_CONFIG_PATH = path.join(REGISTRY_DIR, 'sync-config.json');
const PROJECTS_REGISTRY_PATH = path.join(REGISTRY_DIR, 'projects-registry.json');
const OUTPUT_PATH = path.join(REGISTRY_DIR, 'hybrid-registry.json');

/**
 * Detect local dependencies in skill content.
 * Pattern from discover-skills.py lines 119-138.
 *
 * @param {string} content - SKILL.md content
 * @param {string} skillDir - Skill directory path
 * @returns {Array<string>} List of relative file paths
 */
function detectDependencies(content, skillDir) {
    const deps = [];

    // Match markdown links: [text](path)
    const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
    let match;

    while ((match = linkPattern.exec(content)) !== null) {
        const linkPath = match[2];

        // Skip external URLs
        if (linkPath.startsWith('http://') || linkPath.startsWith('https://')) {
            continue;
        }

        // Skip anchors
        if (linkPath.startsWith('#')) {
            continue;
        }

        // Resolve relative path
        const fullPath = path.join(skillDir, linkPath);
        const relativePath = path.relative(skillDir, fullPath);

        if (relativePath && !relativePath.startsWith('..')) {
            // Normalize to forward slashes for cross-platform consistency
            deps.push(relativePath.replace(/\\/g, '/'));
        }
    }

    return deps;
}

/**
 * Load existing registry or return empty structure.
 *
 * @returns {Promise<object>} Registry data
 */
async function loadRegistry() {
    const existing = await readJson(OUTPUT_PATH);
    if (existing) {
        return existing;
    }

    return {
        version: '2.0.0',
        last_indexed: new Date().toISOString(),
        sources: [],
        skills: []
    };
}

/**
 * Load projects registry.
 *
 * @returns {Promise<object|null>} Projects registry or null
 */
async function loadProjectsRegistry() {
    return await readJson(PROJECTS_REGISTRY_PATH);
}

/**
 * Load sync configuration.
 *
 * @returns {Promise<object>} Sync config
 */
async function loadSyncConfig() {
    const config = await readJson(SYNC_CONFIG_PATH);
    if (!config) {
        throw new Error(`Sync config not found: ${SYNC_CONFIG_PATH}`);
    }
    return config;
}

/**
 * Find project skill sources from current project.
 * Pattern from discover-skills.py lines 151+.
 *
 * @param {object} syncConfig - Sync configuration
 * @param {object} projectsRegistry - Projects registry
 * @returns {Promise<Array<object>>} List of project sources with metadata
 */
async function findProjectSkillSources(syncConfig, projectsRegistry) {
    const sources = [];

    // Check if current directory is a registered project
    const cwd = process.cwd();

    if (!projectsRegistry || !projectsRegistry.projects) {
        return sources;
    }

    // Find project matching current directory
    for (const [projectPath, projectInfo] of Object.entries(projectsRegistry.projects)) {
        const normalizedProjectPath = path.normalize(projectPath);
        const normalizedCwd = path.normalize(cwd);

        // Check if CWD is within this project
        if (normalizedCwd.startsWith(normalizedProjectPath)) {
            // Look for .claude/skills directory
            const skillsDir = path.join(projectPath, '.claude', 'skills');

            if (await fileExists(skillsDir)) {
                sources.push({
                    type: 'project',
                    path: skillsDir,
                    priority: 2,
                    project_name: projectInfo.name || path.basename(projectPath),
                    project_path: projectPath
                });
            }
        }
    }

    // Also scan project_sources from sync-config
    if (syncConfig.project_sources) {
        for (const sourceConfig of syncConfig.project_sources) {
            // Handle both string patterns and object configs
            const projectPattern = typeof sourceConfig === 'string'
                ? sourceConfig
                : sourceConfig.project_pattern;
            const skillsPath = typeof sourceConfig === 'string'
                ? '.claude/skills'
                : (sourceConfig.skills_path || '.claude/skills');
            const priority = typeof sourceConfig === 'string'
                ? 2
                : (sourceConfig.priority || 2);

            // DON'T use expandPath on glob patterns - it breaks them!
            // expandPath('**/foo') => 'C:\...\**\foo' which is invalid

            // Use glob to find directories matching pattern
            // Search from common project roots, not just CWD
            const searchRoots = [
                path.join(os.homedir(), 'OneDrive', 'Coding', '_Projets de code'),
                path.join(os.homedir(), 'Projects'),
                path.join(os.homedir(), 'Code'),
                process.cwd()
            ];

            let matches = [];
            for (const root of searchRoots) {
                if (await fileExists(root)) {
                    const found = await fg(projectPattern, {
                        onlyDirectories: true,
                        absolute: true,
                        ignore: ['**/node_modules/**', '**/.git/**'],
                        cwd: root
                    });
                    matches.push(...found);
                }
            }
            // Dedupe matches
            matches = [...new Set(matches)];

            for (const matchPath of matches) {
                // Build skills directory path
                const skillsDir = path.join(matchPath, skillsPath);

                if (await fileExists(skillsDir)) {
                    sources.push({
                        type: 'project',
                        path: skillsDir,
                        priority,
                        project_name: path.basename(matchPath),
                        project_path: matchPath
                    });
                }
            }
        }
    }

    return sources;
}

/**
 * Scan a source directory for SKILL.md files.
 * Pattern from discover-skills.py lines 227-276.
 *
 * @param {object} source - Source metadata {type, path, priority}
 * @returns {Promise<Array<object>>} List of discovered skills
 */
async function scanSource(source) {
    const skills = [];

    const { type, path: sourcePath, priority } = source;

    console.log(`Scanning ${type} source: ${sourcePath}`);

    if (!await fileExists(sourcePath)) {
        console.warn(`Source path does not exist: ${sourcePath}`);
        return skills;
    }

    // Find all SKILL.md files
    const skillFiles = await fg('**/SKILL.md', {
        cwd: sourcePath,
        absolute: true,
        ignore: ['**/node_modules/**', '**/.git/**', '**/deprecated/**', '**/assets/**', '**/templates/**', '**/examples/**', '**/backups/**']
    });

    console.log(`Found ${skillFiles.length} SKILL.md files in ${type} source`);

    for (const skillFile of skillFiles) {
        try {
            const content = await readFile(skillFile);
            const yaml = extractYaml(content);

            if (!yaml.name) {
                console.warn(`Skipping ${skillFile}: no name in frontmatter`);
                continue;
            }

            const skillDir = path.dirname(skillFile);
            const skillName = yaml.name;

            // Extract metadata
            const description = yaml.description || '';
            const triggers = yaml.triggers || [];
            const allowedTools = yaml['allowed-tools'] || [];
            const license = yaml.license || 'Unknown';
            const metadata = yaml.metadata || {};

            // Compute content hash
            const contentHash = computeHash(content);

            // Detect dependencies
            const dependencies = detectDependencies(content, skillDir);

            // Extract content summary
            const summary = extractContentSummary(content);

            // Build skill entry
            const skill = {
                name: skillName,
                description,
                source: type,
                priority,
                path: skillDir,
                skill_file: skillFile,
                triggers,
                allowed_tools: allowedTools,
                license,
                metadata,
                content_hash: contentHash,
                dependencies,
                summary,
                discovered_at: new Date().toISOString()
            };

            // Add project context if applicable
            if (source.project_name) {
                skill.project_name = source.project_name;
                skill.project_path = source.project_path;
            }

            skills.push(skill);

        } catch (error) {
            console.error(`Error processing ${skillFile}:`, error.message);
        }
    }

    return skills;
}

/**
 * Resolve skill conflicts based on priority.
 * Pattern from discover-skills.py lines 279-313.
 *
 * When multiple skills have the same name, keep the one with highest priority.
 * Priority: Project (2) > Global (1) > Marketplace (0)
 *
 * @param {Array<object>} allSkills - All discovered skills
 * @returns {Array<object>} Resolved skills (duplicates removed)
 */
function resolveSkills(allSkills) {
    const skillMap = new Map();

    for (const skill of allSkills) {
        const name = skill.name;

        if (!skillMap.has(name)) {
            skillMap.set(name, skill);
        } else {
            const existing = skillMap.get(name);

            // Replace if new skill has higher priority
            if (skill.priority > existing.priority) {
                console.log(`Priority override: ${name} (${skill.source} priority ${skill.priority} > ${existing.source} priority ${existing.priority})`);
                skillMap.set(name, skill);
            } else if (skill.priority === existing.priority) {
                console.warn(`Duplicate skill at same priority: ${name} (keeping first found)`);
            }
        }
    }

    return Array.from(skillMap.values());
}

/**
 * Main discovery function.
 *
 * @param {object} options - CLI options
 * @returns {Promise<object>} Registry data
 */
async function main(options) {
    console.log('Starting skill discovery...');
    console.log(`Marketplace root: ${MARKETPLACE_ROOT}`);
    console.log(`Claude home: ${CLAUDE_HOME}`);

    // Load configurations
    const syncConfig = await loadSyncConfig();
    const projectsRegistry = await loadProjectsRegistry();

    // Define sources
    const sources = [
        {
            type: 'marketplace',
            path: path.join(MARKETPLACE_ROOT, 'skills'),
            priority: 0
        },
        {
            type: 'global',
            path: path.join(CLAUDE_HOME, 'skills'),
            priority: 1
        }
    ];

    // Add project sources
    const projectSources = await findProjectSkillSources(syncConfig, projectsRegistry);
    sources.push(...projectSources);

    console.log(`\nScanning ${sources.length} sources:`);
    for (const source of sources) {
        console.log(`  - ${source.type} (priority ${source.priority}): ${source.path}`);
    }

    // Scan all sources
    let allSkills = [];

    for (const source of sources) {
        const skills = await scanSource(source);
        allSkills.push(...skills);
    }

    console.log(`\nTotal skills discovered: ${allSkills.length}`);

    // Resolve conflicts
    const resolvedSkills = resolveSkills(allSkills);

    console.log(`Skills after priority resolution: ${resolvedSkills.length}`);

    // Build registry
    const registry = {
        version: '2.0.0',
        last_indexed: new Date().toISOString(),
        sources: sources.map(s => ({
            type: s.type,
            path: s.path,
            priority: s.priority,
            project_name: s.project_name
        })),
        skills: resolvedSkills,
        stats: {
            total_skills: resolvedSkills.length,
            by_source: {
                marketplace: resolvedSkills.filter(s => s.source === 'marketplace').length,
                global: resolvedSkills.filter(s => s.source === 'global').length,
                project: resolvedSkills.filter(s => s.source === 'project').length
            },
            with_triggers: resolvedSkills.filter(s => s.triggers.length > 0).length,
            with_dependencies: resolvedSkills.filter(s => s.dependencies.length > 0).length
        }
    };

    // Write output
    if (!options.dryRun) {
        await writeJson(OUTPUT_PATH, registry);
        console.log(`\nRegistry written to: ${OUTPUT_PATH}`);
    } else {
        console.log('\n[DRY RUN] Registry not written');
    }

    // Print summary
    console.log('\n=== SUMMARY ===');
    console.log(`Total skills: ${registry.stats.total_skills}`);
    console.log(`  - Marketplace: ${registry.stats.by_source.marketplace}`);
    console.log(`  - Global: ${registry.stats.by_source.global}`);
    console.log(`  - Project: ${registry.stats.by_source.project}`);
    console.log(`With triggers: ${registry.stats.with_triggers}`);
    console.log(`With dependencies: ${registry.stats.with_dependencies}`);

    return registry;
}

// CLI
/* istanbul ignore next */
if (require.main === module) {
    program
        .name('discover-skills')
        .description('Discover and index skills from all sources')
        .option('--dry-run', 'Run without writing output')
        .option('--json', 'Output JSON to stdout')
        .parse(process.argv);

    const options = program.opts();

    main(options)
        .then(registry => {
            if (options.json) {
                console.log(JSON.stringify(registry, null, 2));
            }
            process.exit(0);
        })
        .catch(error => {
            console.error('Error:', error.message);
            console.error(error.stack);
            process.exit(1);
        });
}

module.exports = {
    main,
    scanSource,
    resolveSkills,
    detectDependencies,
    loadRegistry,
    loadProjectsRegistry,
    loadSyncConfig,
    findProjectSkillSources
};
